### ISSUE TITLE
Invalid number of channels and messages in nchan_stub_status

### ISSUE THREAD

**Issue by PhilETaylor** [slact/nchan#453]
```
total published messages: 1512517
stored messages: 18446744073709551556
shared memory used: 200K
channels: 27
subscribers: 25
redis pending commands: 0
redis connected servers: 0
total interprocess alerts received: 6587469
interprocess alerts in transit: 2
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 0
```

Every now again when I run a **nchan_stub_status** I get this impossibly huge number starting 1844... 

E.g. Above you can see 18446744073709551556

Sometimes its the number of channels, sometimes its the number of subscribers sometimes its the stored messages - sometimes its all of these, 

As you can see, with only ever 1512517published, this 18446744073709551556 is impossible, and after a few refreshes/time it reverts back to normal... 

In the time taken to write this issue, its gone again, this is now the output

```
total published messages: 1514222
stored messages: 7
shared memory used: 256K
channels: 28
subscribers: 24
redis pending commands: 0
redis connected servers: 0
total interprocess alerts received: 6595081
interprocess alerts in transit: 2
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 0
```

**Comment by neben** [comment]
We get the same error:
```
total published messages: 15
stored messages: 18446744073709551592
shared memory used: 28K
channels: 18446744073709551596
subscribers: 1
redis pending commands: 0
redis connected servers: 0
total interprocess alerts received: 182
interprocess alerts in transit: 2
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 0
```

Our config:
```
	location = /pubsub {
		nchan_pubsub;
		nchan_subscriber_channel_id $arg_sub;
		nchan_publisher_channel_id $arg_pub;
		nchan_channel_id_split_delimiter ",";
		nchan_message_buffer_length 3;
		nchan_message_timeout 24h;
	}

	location /nchan_stub_status {
		nchan_stub_status;
		add_header 'Access-Control-Allow-Origin' '$http_origin' always;
		add_header 'Access-Control-Allow-Methods' 'GET' always;
	}
```

**Comment by slact** [comment]
Had there been any worker crashes prior to this? I'd like to look a the error log from the start of the nginx run until you notice this issue.

**Comment by PhilETaylor** [comment]
Nope - I pretty much noticed as soon as I started using nchan for the very first time, with less than 100 messages through it since start.

**Comment by PhilETaylor** [comment]
right now my stats are

```
total published messages: 21081434
stored messages: 141
shared memory used: 228K
channels: 49
subscribers: 33
redis pending commands: 0
redis connected servers: 0
total interprocess alerts received: 92090295
interprocess alerts in transit: 0
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 44
```

which tells you nothing other than this bug is intermittent

**Comment by PhilETaylor** [comment]
apparently `184467440737095515` (the prefix to the correct number) is the maximum value of an int64 type...  does that help ?

**Comment by slact** [comment]
@PhilETaylor yep, I'm aware of that. Basically the number of channels or stored messages is negative. The question of "why" still remains.

> which tells you nothing other than this bug is intermittent

Does that mean that the value fixes itself after some time? If so, how long does it take?

**Comment by PhilETaylor** [comment]
I'll keep a closer eye on it for you so I can formulate an answer to that question specifically, but yes, the problem comes and goes with zero intervention from me, and without restarting nginx.

**Comment by PhilETaylor** [comment]
I have just updated everything to the latest versions available, and restated nginx, and published the stats to the following URL - feel free to set a monitor to watch those stats, and alert when it happens. 

https://manage.myjoomla.com/websocketstats/

```
total published messages: 3731
stored messages: 37
shared memory used: 184K
channels: 20
subscribers: 14
redis pending commands: 0
redis connected servers: 0
total interprocess alerts received: 18145
interprocess alerts in transit: 8
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 0
```

**Comment by PhilETaylor** [comment]
currently 

```
total published messages: 619944
stored messages: 18446744073709551590
shared memory used: 40K
channels: 10
subscribers: 9
redis pending commands: 0
redis connected servers: 0
total interprocess alerts received: 2851302
interprocess alerts in transit: 8
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 0
```

**Comment by slact** [comment]
can you post your error log?

**Comment by PhilETaylor** [comment]
The nginx error.log is empty :(

**Comment by slact** [comment]
is it a root-level error log or location-level? A root-level log may contain a lot more information.

**Comment by PhilETaylor** [comment]
I only had the default log setup in nginx.conf, I have set up one in the location, and the server blocks now just to check - I'll check their content tomorrow

**Comment by slact** [comment]
No, it should be _outside_ the server blocks, where you declare the number or workers.

**Comment by PhilETaylor** [comment]
nginx.conf is pretty standard

```
http {

	##
	# Basic Settings
	##

	sendfile on;
	tcp_nopush on;
	tcp_nodelay on;
	keepalive_timeout 65;
	types_hash_max_size 2048;
	# server_tokens off;

	# server_names_hash_bucket_size 64;
	# server_name_in_redirect off;

	include /etc/nginx/mime.types;
	default_type application/octet-stream;

	##
	# SSL Settings
	##

	ssl_protocols TLSv1 TLSv1.1 TLSv1.2; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;

	##
	# Logging Settings
	##

	access_log /var/log/nginx/access.log;
	error_log /var/log/nginx/error.log;

	##
	# Gzip Settings
	##

	gzip on;

	# gzip_vary on;
	# gzip_proxied any;
	# gzip_comp_level 6;
	# gzip_buffers 16 8k;
	# gzip_http_version 1.1;
	# gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;

	##
	# Virtual Host Configs
	##

	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;
}
```

**Comment by slact** [comment]
It needs to be: 
```
error_log /var/log/nginx/error.log notice;

http {

	##
	# Basic Settings
	##

	sendfile on;
	tcp_nopush on;
	tcp_nodelay on;
	keepalive_timeout 65;
	types_hash_max_size 2048;
	# server_tokens off;

	# server_names_hash_bucket_size 64;
	# server_name_in_redirect off;

	include /etc/nginx/mime.types;
	default_type application/octet-stream;

	##
	# SSL Settings
	##

	ssl_protocols TLSv1 TLSv1.1 TLSv1.2; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;

	##
	# Logging Settings
	##

	access_log /var/log/nginx/access.log;

	##
	# Gzip Settings
	##

	gzip on;

	# gzip_vary on;
	# gzip_proxied any;
	# gzip_comp_level 6;
	# gzip_buffers 16 8k;
	# gzip_http_version 1.1;
	# gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;

	##
	# Virtual Host Configs
	##

	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;
}
```

**Comment by PhilETaylor** [comment]
```
total published messages: 198
stored messages: 18446744073709551553
shared memory used: 80K
channels: 18446744073709551596
subscribers: 0
redis pending commands: 0
redis connected servers: 0
total interprocess alerts received: 737
interprocess alerts in transit: 32
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 0
```

```
2018/06/29 08:40:21 [alert] 27750#27750: *1004393 open socket #21 left in connection 36
2018/06/29 08:40:21 [alert] 27750#27750: *1006328 open socket #26 left in connection 44
2018/06/29 08:40:21 [alert] 27750#27750: aborting
```

**Comment by slact** [comment]
Ok, that helps. Would you be willing to run a version of Nchan that intentionally crashes if the totals ever get weird so I can look at the coredump?

**Comment by PhilETaylor** [comment]
This is actually a live SaaS and so any downtime is not really possible. Especially to debug something thats not actually breaking its use. 

The log so far

```
2018/06/29 08:40:18 [info] 3441#3441: Using 32768KiB of shared memory for nchan in /etc/nginx/nginx.conf:64
2018/06/29 08:40:21 [notice] 3455#3455: signal process started
2018/06/29 08:40:21 [notice] 23592#23592: using the "epoll" event method
2018/06/29 08:40:21 [notice] 23592#23592: start worker processes
2018/06/29 08:40:21 [notice] 23592#23592: start worker process 3460
2018/06/29 08:40:21 [notice] 23592#23592: start worker process 3461
2018/06/29 08:40:21 [notice] 23592#23592: start worker process 3462
2018/06/29 08:40:21 [alert] 27749#27749: *716123 open socket #21 left in connection 9
2018/06/29 08:40:21 [alert] 27749#27749: *1016018 open socket #24 left in connection 10
2018/06/29 08:40:21 [alert] 27749#27749: *1016393 open socket #20 left in connection 11
2018/06/29 08:40:21 [alert] 27749#27749: *989035 open socket #33 left in connection 30
2018/06/29 08:40:21 [alert] 27749#27749: aborting
2018/06/29 08:40:21 [alert] 27748#27748: *810731 open socket #26 left in connection 12
2018/06/29 08:40:21 [alert] 27748#27748: *183123 open socket #35 left in connection 13
2018/06/29 08:40:21 [alert] 27748#27748: *1017615 open socket #25 left in connection 17
2018/06/29 08:40:21 [alert] 27748#27748: *117676 open socket #21 left in connection 18
2018/06/29 08:40:21 [alert] 27748#27748: *863030 open socket #3 left in connection 19
2018/06/29 08:40:21 [alert] 27748#27748: *1016896 open socket #24 left in connection 34
2018/06/29 08:40:21 [alert] 27748#27748: aborting
2018/06/29 08:40:21 [alert] 27750#27750: *947061 open socket #3 left in connection 19
2018/06/29 08:40:21 [alert] 27750#27750: *1012297 open socket #39 left in connection 22
2018/06/29 08:40:21 [alert] 27750#27750: *1004393 open socket #21 left in connection 36
2018/06/29 08:40:21 [alert] 27750#27750: *1006328 open socket #26 left in connection 44
2018/06/29 08:40:21 [alert] 27750#27750: aborting
2018/06/29 08:40:21 [notice] 23592#23592: signal 17 (SIGCHLD) received from 27750
2018/06/29 08:40:21 [notice] 23592#23592: worker process 27748 exited with code 0
2018/06/29 08:40:21 [notice] 23592#23592: worker process 27750 exited with code 0
2018/06/29 08:40:21 [notice] 23592#23592: signal 29 (SIGIO) received
2018/06/29 08:40:21 [notice] 23592#23592: signal 17 (SIGCHLD) received from 27749
2018/06/29 08:40:21 [notice] 23592#23592: worker process 27749 exited with code 0
2018/06/29 08:40:21 [notice] 23592#23592: signal 29 (SIGIO) received
2018/06/29 09:34:51 [notice] 12906#12906: signal process started
2018/06/29 09:34:51 [notice] 23592#23592: signal 1 (SIGHUP) received from 12906, reconfiguring
2018/06/29 09:34:51 [notice] 23592#23592: reconfiguring
2018/06/29 09:34:51 [notice] 23592#23592: using the "epoll" event method
2018/06/29 09:34:51 [notice] 23592#23592: start worker processes
2018/06/29 09:34:51 [notice] 23592#23592: start worker process 12909
2018/06/29 09:34:51 [notice] 23592#23592: start worker process 12910
2018/06/29 09:34:51 [notice] 23592#23592: start worker process 12912
2018/06/29 09:34:51 [notice] 3460#3460: gracefully shutting down
2018/06/29 09:34:51 [notice] 3461#3461: gracefully shutting down
2018/06/29 09:34:51 [notice] 3460#3460: exiting
2018/06/29 09:34:51 [notice] 3461#3461: exiting
2018/06/29 09:34:51 [notice] 3460#3460: nchan: 14 channels remain at exit
2018/06/29 09:34:51 [notice] 3461#3461: nchan: 18 channels remain at exit
2018/06/29 09:34:51 [notice] 3460#3460: nchan: 16 unbuffered messages remain at exit
2018/06/29 09:34:51 [notice] 3461#3461: nchan: 110 unbuffered messages remain at exit
2018/06/29 09:34:51 [alert] 3460#3460: *1038539 open socket #25 left in connection 9
2018/06/29 09:34:51 [alert] 3460#3460: *1038573 open socket #26 left in connection 10
2018/06/29 09:34:51 [alert] 3460#3460: aborting
2018/06/29 09:34:51 [notice] 3462#3462: gracefully shutting down
2018/06/29 09:34:51 [notice] 3462#3462: exiting
2018/06/29 09:34:51 [notice] 3462#3462: nchan: 18 channels remain at exit
2018/06/29 09:34:51 [notice] 3462#3462: nchan: 39 unbuffered messages remain at exit
2018/06/29 09:34:51 [notice] 3460#3460: exit
2018/06/29 09:34:51 [notice] 3462#3462: exit
2018/06/29 09:34:51 [notice] 23592#23592: signal 17 (SIGCHLD) received from 3460
2018/06/29 09:34:51 [notice] 23592#23592: worker process 3460 exited with code 0
2018/06/29 09:34:51 [notice] 23592#23592: signal 29 (SIGIO) received
2018/06/29 09:34:51 [alert] 3461#3461: *1040562 open socket #33 left in connection 10
2018/06/29 09:34:51 [alert] 3461#3461: *1038623 open socket #4 left in connection 11
2018/06/29 09:34:51 [alert] 3461#3461: *1039635 open socket #21 left in connection 13
2018/06/29 09:34:51 [alert] 3461#3461: aborting
2018/06/29 09:34:51 [notice] 3461#3461: exit
2018/06/29 09:34:51 [notice] 23592#23592: signal 17 (SIGCHLD) received from 3462
2018/06/29 09:34:51 [notice] 23592#23592: worker process 3461 exited with code 0
2018/06/29 09:34:51 [notice] 23592#23592: worker process 3462 exited with code 0
2018/06/29 09:34:51 [notice] 23592#23592: signal 29 (SIGIO) received
2018/06/29 09:35:01 [notice] 23592#23592: signal 3 (SIGQUIT) received from 12985, shutting down
2018/06/29 09:35:01 [notice] 12910#12910: gracefully shutting down
2018/06/29 09:35:01 [notice] 12910#12910: exiting
2018/06/29 09:35:01 [notice] 12909#12909: gracefully shutting down
2018/06/29 09:35:01 [notice] 12909#12909: exiting
2018/06/29 09:35:01 [notice] 12910#12910: nchan: 14 channels remain at exit
2018/06/29 09:35:01 [notice] 12910#12910: nchan: 92 unbuffered messages remain at exit
2018/06/29 09:35:01 [notice] 12909#12909: nchan: 11 channels remain at exit
2018/06/29 09:35:01 [notice] 12909#12909: nchan: 14 unbuffered messages remain at exit
2018/06/29 09:35:01 [notice] 12912#12912: gracefully shutting down
2018/06/29 09:35:01 [notice] 12912#12912: exiting
2018/06/29 09:35:01 [notice] 12912#12912: nchan: 11 channels remain at exit
2018/06/29 09:35:01 [notice] 12912#12912: nchan: 42 unbuffered messages remain at exit
2018/06/29 09:35:01 [alert] 12912#12912: *1040755 open socket #3 left in connection 9
2018/06/29 09:35:01 [alert] 12912#12912: aborting
2018/06/29 09:35:01 [notice] 12912#12912: exit
2018/06/29 09:35:01 [notice] 12909#12909: exit
2018/06/29 09:35:01 [notice] 23592#23592: signal 17 (SIGCHLD) received from 12909
2018/06/29 09:35:01 [error] 12910#12910: MEMSTORE:04: force-reaping msg with refcount 1
2018/06/29 09:35:01 [notice] 23592#23592: worker process 12909 exited with code 0
2018/06/29 09:35:01 [notice] 23592#23592: worker process 12912 exited with code 0
2018/06/29 09:35:01 [notice] 23592#23592: signal 29 (SIGIO) received
2018/06/29 09:35:01 [notice] 23592#23592: signal 17 (SIGCHLD) received from 12912
2018/06/29 09:35:01 [notice] 12910#12910: exit
2018/06/29 09:35:01 [notice] 23592#23592: signal 17 (SIGCHLD) received from 12910
2018/06/29 09:35:01 [notice] 23592#23592: worker process 12910 exited with code 0
2018/06/29 09:35:01 [notice] 23592#23592: exit
2018/06/29 09:35:01 [info] 12990#12990: Using 32768KiB of shared memory for nchan in /etc/nginx/nginx.conf:64
2018/06/29 09:35:01 [notice] 12996#12996: using the "epoll" event method
2018/06/29 09:35:01 [notice] 12996#12996: nginx/1.15.0
2018/06/29 09:35:01 [notice] 12996#12996: OS: Linux 4.4.0-127-generic
2018/06/29 09:35:01 [notice] 12996#12996: getrlimit(RLIMIT_NOFILE): 1024:4096
2018/06/29 09:35:01 [notice] 12999#12999: start worker processes
2018/06/29 09:35:01 [notice] 12999#12999: start worker process 13000
2018/06/29 09:35:01 [notice] 12999#12999: start worker process 13001
2018/06/29 09:35:01 [notice] 12999#12999: start worker process 13002
2018/06/30 06:25:05 [notice] 12999#12999: signal 10 (SIGUSR1) received from 3261, reopening logs
2018/06/30 06:25:05 [notice] 12999#12999: reopening logs
2018/06/30 06:25:05 [notice] 13002#13002: reopening logs
2018/06/30 06:25:05 [notice] 13000#13000: reopening logs
2018/06/30 06:25:05 [notice] 13001#13001: reopening logs
```

**Comment by slact** [comment]
Understandable. I'll keep digging.

**Comment by slact** [comment]
Stats could have gotten corrupted after an Nginx worker crash. fixed in 1.3.x
